Prometheus: Metric encyclopedia ux collab design

bohandley opened the pull request:
**What is this feature?**
Q2 design updates for the metric encyclopedia component in the prometheus query builder. These updates are based on [collaboration with @catherineymgui](https://www.figma.com/file/BRH4zzJYD4dZWJw8Ix2OfH/Metrics-Encyclopedia?type=design&node-id=46-41473&t=5UyOh1aPTUwbF8Uu-0) and other feedback from Grafana users documented here: https://docs.google.com/document/d/1DlyklbU2jcOPb4P_s7ugLqBLpFIacH3Rs4BApVTW4Gk/edit#heading=h.dvkx4lxhaidj

**Why do we need this feature?**
This is another iteration in the feedback lifecycle for the metric encyclopedia. This component was developed from research from the [G10 project Getting Started: Prometheus](https://github.com/orgs/grafana/projects/112/views/22?filterQuery=getting+started&pane=issue&itemId=13178536). Users wanted to know more about their metrics, to be able to search all their metrics, and to explore them in a more meaningful way. The metric encyclopedia allows a user to explore all their metrics, is performant, has a flexible fuzzy search, and allows users to explore metrics by type and description. We first received UX feedback from the afternoon session, then collaborated in the hackathon, received feedback from Grafana users, and are on the next iteration before testing with full story and users outside of Grafana.

**Who is this feature for?**
Beginning and intermediate prometheus users.

Notable updates:
The metric encyclopedia supports accessibility for opening the modal in the metric select and tabbing through table results in the modal. First, when a user opens the metric select, they can use keyboard input to open the modal. After opening the modal and a user types characters to search for a metric name, the user can tab into the table, and select a metric by pressing the enter key. When the search input is focused, a user can press up and down to navigate through metric names and select by pressing enter.

https://github.com/grafana/grafana/assets/25674746/1be64033-076d-4aae-8d15-b12e141c508e

Please check that:
- [x] It works as expected from a user's perspective.
- [x] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

bohandley committed:
add class for full story click event on open modal

bohandley committed:
move feedback link to modal top under header

bohandley committed:
move results amount to bottom left

bohandley committed:
move settings into modal, change language from exclude to include

bohandley committed:
add metadata to backend search, use toggletip for settings, clean code

bohandley committed:
style input row, remove labels and update settings button design

bohandley committed:
remove alphabet search as requested by design

bohandley committed:
display selected metric

bohandley committed:
update style warning message for labels filtered metrics

bohandley committed:
organize results footer

bohandley committed:
update table design w fixed width and sticky header

bohandley committed:
allow focus row on tab and use key Enter to select metric on keydown

bohandley committed:
add rudderstack event for disable text wrap

bohandley committed:
add messages for no metrics found, labels, search and none in data source.

bohandley committed:
filter by type placeholder

catherineymgui commented:
Looking good! 

Things I've noticed:
- [ ] UI bug when I select a label filter
<img width="717" alt="Screenshot 2023-05-15 at 14 01 35" src="https://github.com/grafana/grafana/assets/5264102/908a326d-46f7-4d25-875c-756d31779657">
the select dropdown becomes super small
<img width="1022" alt="Screenshot 2023-05-15 at 14 04 39" src="https://github.com/grafana/grafana/assets/5264102/6dacf12f-d362-46b8-bf86-f8d7071505a2">
the popup modal has a scroll bar showing up for the whole modal

- [ ] Should there be an overflow-wrap or smth when the metric name is too long? 
<img width="645" alt="Screenshot 2023-05-15 at 15 02 47" src="https://github.com/grafana/grafana/assets/5264102/b4848f2c-43cc-4762-a174-30d68cdd55d4">

- [ ] Have 4px margin b/w the search row and the “currently selected” text. 16px between “currently selected text” and the table 
- [ ] Add some padding inside the first table header row (8 pixels on all sides)
- [ ] Are we able to use less padding for the toggletip / additional settings (8px top and bottom, 16px on the right and left)
- [ ] Additional settings dropdown — what is the font size on these? It should be 12px but it looks smaller and I cannot inspect it 
- [ ] Additional settings dropdown - 4px padding between each option
- [ ] 24px margin between the last table cell and the pagination row
- [ ] # of Results per page — have this also be the same font size as the “Showing x results” (0.85rem?)
- [ ] 8px margin b/w the “# of results per page” and the dropdown

bohandley committed:
add min width to custom select option

bohandley committed:
add text wrap for long metric names

bohandley committed:
Have 4px margin b/w the search row and the 'currently selected' text. 16px between 'currently selected text' and the table

bohandley committed:
Add some padding inside the first table header row (8 pixels on all sides)

bohandley committed:
font-size of 12px for additional settings text

bohandley committed:
4px padding between additional settings text

bohandley committed:
24px margin between the last table cell and the pagination row

bohandley committed:
# of Results per page font size 0.85rem

bohandley committed:
8px margin b/w the '# of results per page' and the dropdown

bohandley committed:
Merge branch 'main' into bohandley/prometheus-metric-encyclopedia-ux-collab

bohandley committed:
fix test

bohandley committed:
Merge branch 'bohandley/prometheus-metric-encyclopedia-ux-collab' of https://github.com/grafana/grafana into bohandley/prometheus-metric-encyclopedia-ux-collab

bohandley committed:
add infer type setting for testing

bohandley committed:
use title on icon instead of wrapping in tooltip to fix test

bohandley committed:
fix icon issue

bohandley committed:
italicize inferred types, update setting text and add icon

bohandley committed:
add space for label filters alert message

itsmylife commented on public/app/plugins/datasource/prometheus/querybuilder/components/metrics-modal/state/helpers.ts:
```diff
@@ -32,7 +33,16 @@ export async function setMetrics(
 
   metricsData = initialMetrics?.map((m: string) => {
-    const type = getMetadataType(m, datasource.languageProvider.metricsMetadata!);
+    let type = getMetadataType(m, datasource.languageProvider.metricsMetadata!);
+    let inferredType = false;
+    if (!type && inferType) {
+      type = metricTypeHints(m);
+
+      if (type) {
+        inferredType = true;
+      }
+    }
```
In this code block, wouldn't the `inferredType` always be false? I think it should have been defined outside of the `map` callback. Or am I wrong?

itsmylife replied:
Same code is in `helpers.ts` too.

bohandley replied:
```
    let type = getMetadataType(m, datasource.languageProvider.metricsMetadata!);
    // inferredType will be set on each metric data element, and it will always be false unless the setting (inferType) is turned on and a type has been inferred
    let inferredType = false;
    // inferType is the setting and if turned on, then we check each metric and update the inferredType in the metric data
    if (!type && inferType) {
      type = metricTypeHints(m);

      if (type) {
        inferredType = true;
      }
    }
```

bohandley replied:
I can add some comments to this code to clarify what is happening. 

`inferType` is the state setting to use the inferring logic in `metricTypeHints`
If it is turned off, each metric data element will be given the inferredType field of false.

If `inferType` setting is true, then we perform the logic for each metric. 
If we set `inferredType` outside the map, if a metric does not have a type, if then inferred a type, the flag is set true, and the next metric does have a type, the flag will still be set true and that metric with an actual type will be given a field that it was inferred. 

So, we need to set that default to false for each metric inside the map.

itsmylife replied:
Yeah, you are right I missed some part of the code.

itsmylife commented:
I see the same thing like @catherineymgui pointed:
<img width="1020" alt="image" src="https://github.com/grafana/grafana/assets/820946/9fbb4e52-0a0a-465e-a35a-b54264415cfb">

itsmylife commented:
I think histograms are marked as counters 
<img width="1020" alt="image" src="https://github.com/grafana/grafana/assets/820946/7c26c9d4-dc3e-4c56-b7ed-2d48cf8f0029">

bohandley committed:
make open button style consistent with advanced datasource picker

bohandley committed:
keep copy for open modal button

bohandley committed:
refactor rudderstack interactions and add inferType

bohandley committed:
add event tracking for opening the modal

gtk-grafana commented:
Horizontal scrollbar in metrics is new:
<img width="533" alt="image" src="https://github.com/grafana/grafana/assets/109082771/582c63ba-c396-4c9a-8a31-3325a0038f23">

How to tab into metrics catalog?
<img width="524" alt="image" src="https://github.com/grafana/grafana/assets/109082771/f8882fae-7af5-427f-9781-f22f072f65ff">

Bug with max items per page. If you type 10000, it won't display the max number as more then 1000, but it will show more than 1000 items per page.
<img width="1049" alt="image" src="https://github.com/grafana/grafana/assets/109082771/ea41b63d-e6a0-417d-986c-d45677b84f70">

bohandley commented:
> I see the same thing like @catherineymgui pointed: <img alt="image" width="1020" src="https://user-images.githubusercontent.com/820946/240589962-9fbb4e52-0a0a-465e-a35a-b54264415cfb.png">

Can you clarify what do you see @itsmylife? @catherineymgui point was that there was a scrollbar when there was only one metric.

bohandley commented:
> I think histograms are marked as counters <img alt="image" width="1020" src="https://user-images.githubusercontent.com/820946/240590698-7c26c9d4-dc3e-4c56-b7ed-2d48cf8f0029.png">

Ok, there's a few things wrapped up in this. While a histogram is a type, it is also a function that creates other types. 

`grafana_http_request_duration_seconds` is the histogram which creates the following types:
`grafana_http_request_duration_seconds_bucket` = counter type
`grafana_http_request_duration_seconds_count` = counter type
`grafana_http_request_duration_seconds_sum` = counter type

In our query builder hints we make this even more confusing. [We give the impression that metrics ending in _bucket are a histogram.](https://github.com/grafana/grafana/blob/509281d537c25c54ec03f0091f55577e15897a1c/public/app/plugins/datasource/prometheus/query_hints.ts#L16-L87) We also give an impression that metrics ending in _count, _sum and _total are counters. We can see above that sometimes metrics ending in _count and _sum could be histograms as well, and that they are sometimes created as child metrics out of histograms. The hints are added for operations though, where _sum, _total and _count get the `rate()` hint and _bucket gets the `histogram_quantile()` hint.

I would like to schedule a meeting with some Prometheus folks to see if we need to change the metric type logic of what we show. While prometheus says a metric that ends in _bucket is a counter, it was created as histogram so we should probably add that detail too. 

Maybe we can add (histogram) after things that were created as part of a histogram if 1. it ends in _bucket or 2. The description says it is part of a histogram.

bohandley committed:
galen's feedback, fix select horizontal scroll and results perpg bug

bohandley commented:
@gtk-grafana feedback response here!
1. Removed the select horizontal scroll "feature" 🤣 
2. The tabbing to the button is not supported by the select. This is consistent with the Datasource Dropdown picker. Tab is used to select an element only. The previous M.E. behavior of having the button as a modal open but also an option was nice, it allowed tab to select, but I think matching for consistency is more important at this point in the development of the M.E. I think this is a good case for taking it up with the Saga folk though and starting a discussion about changing the behavior of the select to tab through results instead of selecting an option.
3. Results per page bug fixed 🎉

gtk-grafana commented:
@bohandley Re:
> The tabbing to the button is not supported by the select.
How do users without a mouse use this feature?

bohandley commented:
> @bohandley Re:
> 
> > The tabbing to the button is not supported by the select.
> 
> How do users without a mouse use this feature?
> 
> Edit: I think I muddied the waters by calling it "tabbing" when we're in the select input you navigate through options via the arrow keys, and selecting an option via spacebar.
> 
> Double edit: This might involve changing the design, (or adding an additional entry point into this feature?), but if the design doesn't allow for users with disabilities it needs to be changed anyway IMO

Good point! The implementation with the custom option did allow for users without a mouse to open the model. The model based on the Datasource dropdown does not allow that functionality. I will revert back to the option as button so that the accessibility is supported.

bohandley committed:
ismail's feedback for metric types

bohandley committed:
revert button in option for accessibility(galen) and style button with ghost mode

bohandley committed:
change name to Metrics explorer

bohandley committed:
fix hover/focus styles

itsmylife commented on public/app/plugins/datasource/prometheus/querybuilder/components/MetricSelect.tsx:
```diff
@@ -131,6 +133,7 @@ export function MetricSelect({
       return (
         <div
           {...props.innerProps}
+          ref={props.innerRef}
```
I just wonder why we need this. Could you explain it to me please?

bohandley replied:
The innerRef and innerProps properties carry forward things like the hover and onClick needed by the option. Without them I would have to add custom styles to match the behavior of the select.

itsmylife commented on public/app/plugins/datasource/prometheus/querybuilder/components/metrics-modal/state/helpers.ts:
```diff
@@ -75,12 +90,10 @@ export function displayedMetrics(state: MetricsModalState, dispatch: React.Dispa
- * @param metrics
- * @param skipLetterSearch used to show the alphabet letters as clickable before filtering out letters (needs to be refactored)
+ * Filter the metrics with all the options, fuzzy, type, null metadata
  * @returns
```
```suggestion
```

bohandley replied:
Fixed in https://github.com/grafana/grafana/pull/68421/commits/6a9b56a801077a5abf8598a16617dc981c4e1f08

itsmylife reviewed (approved):
Some nitpicks. But the rest LGTM 🚀

catherineymgui commented:
Noticed a couple of things

1. Hover on the tooltip icon - it takes a couple of seconds for the tooltip to appear (not instantly) and this is the styling of the tooltip. Not sure if this is just on my end tho
<img width="478" alt="Screenshot 2023-05-26 at 09 13 59" src="https://github.com/grafana/grafana/assets/5264102/694c9955-ed6e-4474-919b-db6a85299513">

2. Ideally if there is only a couple of metrics, they wouldn't take full height of the container. It would appear like so (similar to what @itsmylife pointed out) It would just be like shown below. v 
<img width="920" alt="Screenshot 2023-05-26 at 09 33 34" src="https://github.com/grafana/grafana/assets/5264102/263c9254-94a0-42ae-9c66-9e214377ecc8">

bohandley commented on public/app/plugins/datasource/prometheus/querybuilder/components/metrics-modal/ResultsTable.tsx:
```diff
@@ -36,15 +37,7 @@ export function ResultsTable(props: ResultsTableProps) {
-        selectedTypes: state.selectedTypes,
-        letterSearch: state.letterSearch,
-      });
+      tracking('grafana_prom_metric_encycopedia_tracking', state, metric.value);
```
We should have a different label, but for now, I argue that we should keep this one until our testing is complete. This metric has been in use for testing so far so changing it would add a layer of complexity to our testing. I believe we can change it to the correct name once it is GA so that we can differentiate it from the early testing phase.

`grafana_prometheus_metrics_explorer_tracking` for example

bohandley commented on public/app/plugins/datasource/prometheus/querybuilder/components/metrics-modal/state/helpers.ts:
```diff
@@ -174,13 +192,70 @@ export async function getBackendSearchMetrics(
   return await results.then((results) => {
     return results.map((result) => {
-      return {
+      let type = getMetadataType(result.text, datasource.languageProvider.metricsMetadata!);
+      let inferredType;
+      if (!type && inferType) {
+        type = metricTypeHints(result.text);
+
+        if (type) {
+          inferredType = true;
+        }
+      }
+      const description = getMetadataHelp(result.text, datasource.languageProvider.metricsMetadata!);
+
+      if (description?.toLowerCase().includes('histogram') && type !== 'histogram') {
+        type += ' (histogram)';
+      }
+
+      const metricData: MetricData = {
         value: result.text,
+        type: type,
+        description: description,
+        inferred: inferredType,
       };
+
+      return metricData;
```
Refactoring and adding it to a function to build metric data, good catch!

bohandley replied:
Fixed in https://github.com/grafana/grafana/pull/68421/commits/6a9b56a801077a5abf8598a16617dc981c4e1f08

bohandley commented on public/app/plugins/datasource/prometheus/querybuilder/components/metrics-modal/state/helpers.ts:
```diff
@@ -174,13 +192,70 @@ export async function getBackendSearchMetrics(
   });
 }
 
+function metricTypeHints(metric: string): string | undefined {
```
Changing this to return the empty string, this makes sense. The checks I have implemented work against an empty string.

bohandley committed:
ismail's feedbcak, refactor hints, return empty string, remove @return

bohandley committed:
Fix icon hovering: put tooltips back in over titles on icon

bohandley committed:
make results not expand to fill table space and fix width for modal open option button

bohandley commented:
@catherineymgui 

> Noticed a couple of things
> 
> 1. Hover on the tooltip icon - it takes a couple of seconds for the tooltip to appear (not instantly) and this is the styling of the tooltip. Not sure if this is just on my end tho
> 

> 2. Ideally if there is only a couple of metrics, they wouldn't take full height of the container. It would appear like so (similar to what @itsmylife pointed out) It would just be like shown below. v
> 

@catherineymgui and @itsmylife all changes requested are update! Ismail, I responded to your requests in your comments. Catherine, If there are a couple of metrics they no longer take up the height of the container and I fixed the tooltip issue. I had removed them thinking they would fix a bug which I had fixed in a better way and had not added them back in. Good catch!


Uploading Screen Recording 2023-05-26 at 10.39.26 AM.mov…
